_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,606 @@
+(with an additional restriction that operators *must* be global):
+it is only allowed at file level, the type must be defined in the same file and it makes
+the operator available everywhere.
+Note that only the directive itself must be colocated with the type.
+The operator definitions may be imported from a separate file.
+
+#### More Examples
+
+##### Unchecked Counter
+This example shows how operators can be used to bypass the checked arithmetic for types
+specifically defined for cases where those checks are superfluous:
+
+```solidity
+pragma solidity ^0.8.19;
+
+type UncheckedCounter is uint;
+
+using {add as +, lt as <} for UncheckedCounter global;
+
+UncheckedCounter constant ONE = UncheckedCounter.wrap(1);
+
+function add(UncheckedCounter x, UncheckedCounter y) pure returns (UncheckedCounter) {
+    unchecked {
+        return UncheckedCounter.wrap(
+            UncheckedCounter.unwrap(x) +
+            UncheckedCounter.unwrap(y)
+        );
+    }
+}
+
+function lt(UncheckedCounter x, UncheckedCounter y) pure returns (bool) {
+    return UncheckedCounter.unwrap(x) < UncheckedCounter.unwrap(y);
+}
+
+contract C {
+    uint internalCounter = 12;
+
+    function testCounter() public returns (uint) {
+        for (
+            UncheckedCounter i = UncheckedCounter.wrap(12);
+            i < UncheckedCounter.wrap(20);
+            i = i + ONE
+        )
+            ++internalCounter;
+
+        return internalCounter;
+    }
+}
+```
+
+##### Token Conversion
+This is a more complex example that shows multiple aspects of user-defined operators.
+The `TokenConversion` contract holds balances of two different tokens, defined as UDVTs,
+and uses a third type as a common denominator for calculations.
+
+`TokenConversion.sol`
+```solidity
+pragma solidity ^0.8.19;
+
+import {TokenA, TokenB, BaseToken, ZERO, ONE} from "./Tokens.sol";
+
+contract TokenConversion {
+    mapping(address account => TokenA balance) public balanceOfTokenA;
+    mapping(address account => TokenB balance) public balanceOfTokenB;
+
+    BaseToken averageDepositSize = ZERO;
+    BaseToken numberOfDeposits = ZERO;
+
+    function deposit(address receiver, TokenA amount) public {
+        balanceOfTokenA[receiver] = balanceOfTokenA[receiver] + amount;
+        BaseToken depositSize = amount.toBase();
+        averageDepositSize = (numberOfDeposits * averageDepositSize + depositSize) / (numberOfDeposits + ONE);
+        numberOfDeposits = numberOfDeposits + ONE;
+    }
+
+    function averageDepositSizeAsTokenA() public view returns (TokenA) {
+        return averageDepositSize.toA();
+    }
+
+    function averageDepositSizeAsTokenB() public view returns (TokenB) {
+        return averageDepositSize.toB();
+    }
+
+    function convertAToB(TokenA amount) public {
+        address owner = msg.sender;
+        require(balanceOfTokenA[owner] >= amount);
+
+        balanceOfTokenA[owner] = balanceOfTokenA[owner] + -amount;
```

Comment:
@matheusaaguiar
This is a quick and dirty way I used to shoehorn the unary `-` into the example. Maybe we could come up with something better though.

Reply:
I though about that, but does it make sense to have `-`, since the underlying type is `uint`?